.mailmap

```diff
@@ -1,292 +1,387 @@
-# Prevent git from showing duplicate names with commands like "git shortlog"
-# See the manpage of git-shortlog for details.
-# The syntax is:
-# Name that should be used <email that should be used> Bad name <bad email>
+# Before editing this file, run bin/authors_update.py. This will
+# make sure that all authors known to git are added to AUTHORS
+# and inform you about the types of entries needed here.
 #
-# You can skip Bad name if it is the same as the one that should be used, and is unique.
+# Entries in this file are made for two reasons:
 #
-# This file is up-to-date if the command git log --format="%aN <%aE>" | sort -u
-# gives no duplicates.
-Ondřej Čertík <ondrej@certik.cz> Ondrej Certik <ondrej@certik.cz>
+# 1) to make the name and address of an author appear canonically
+# 2) to merge authors in git that are actually the same person
+#
+# To address either of these issues, add an entry to this file like
+#
+#     Canonical Name <Canonical@email> commit name <commit@email>
+#
+# then
+#    1) iteratively re-run bin/authors_update.py and make edits
+#       here until no error is raised by authors_update
+#    2) run bin/mailmap_update.py to check the results
+#    3) as a final check, commit only changes to .mailmap and
+#       re-run bin/authors_update.py to see all the
+#       authors that will be added, e.g.
+#         $ git add .mailmap; git commit -m mailmap
+#         $ git reset --hard; bin/authors_update.py
+#    4) when all names and addresses in AUTHORS appear as
+#       desired and each entry respresents a distint author
+#       and there is no error when running bin/mailmap_update.py
+#       then commit the results together with the .mailmap changes:
+#         $ git reset head~n (if you've made n commits of changes)
+#         $ git commit -am 'Update .mailmap/AUTHORS. Welcome to SymPy!'
```

Comment:
Why is it necessary to do all this commit shenanigans? I usually just fix everything and commit it in one go.

Reply:
I guess it depends whether you do the commit at every release or at every PR-commit. The AUTHORS is a moving target -- if you make changes to .mailmap then there are new conditions that may be detected in the AUTHORS (in terms of name ambiguity). The reference in the OP to the steps I went through in to make this PR give real-life examples of iteration.

Reply:
If I only write the AUTHORS if there are no ambiguous email addresses then I think the directions can be simplified to this:

```
# Then,
#
#    1) iteratively re-run bin/authors_update.py and make edits
#       here until no error is raised by authors_update.py; if there
#       are only warnings (but no further changes are desired) then
#       proceed to step 2.
#    2) run bin/mailmap_update.py to check the results
#    4) if all names and addresses in AUTHORS appear as
#       desired and multiple entries for a single author don't
#       appear then commit the results:
#
#         $ git commit -am 'Update .mailmap/AUTHORS. Welcome to SymPy!'
#       or
#         $ git commit -am 'Add Foo to AUTHORS. Welcome to SymPy!'
```

Reply:
But neither is dependent on the files being committed. You can create self-consistent .mailmap and AUTHORS and only commit once at the end. 

Maybe you are under the (false) impression that .mailmap has to be committed to take affect? It doesn't. Like .gitignore, git just reads the file. Whether or not it is committed or even tracked at all has no bearing.

Reply:
> and only commit once at the end.

Yep -- that's what the new directions advise in step (4) -- should be step (3). If AUTHORS is not written until one is sure that there are no ambiguous email addresses -- as is now implemented -- then the single commit should work after, perhaps, a re-edit of .mailmap in response to the first set of disambiguation entries.

AUTHORS should never be edited manually. Of course you can, but you shouldn't. Any manual manipulation should happen inside the `authors_update.py` script -- if necessary. For example, if two people ever use the same name and dummy address for a commit then there will be no way to tell .mailmap how to resolve this. We have such a near miss in the history: something like 'unknown <david-PC@none>' and 'David <David-PC@none>'. I modified the authors script to detect such near misses, but if the case of the dummy address had been the same and the authors name the same, it would it would be impossible to disambiguate these two authors. Until one of them changed their commit name or address we couldn't get them to be part of AUTHORS as two different entries. We could add them manually to AUTHORS but that would conflict with later attempts to add a .mailmap entry. The best way to handle this case would (I think) be to manually edit the commit to change the author's commit name or address to make it unambiguous.